File: php/src/Canon.php
<?php
namespace Bonemesh;

// BoneMesh restricted-JCS certificate canonicalization (security.md §11.1): the
// exact byte string the mesh root signs. Byte-for-byte identical to the Java,
// Elixir, Rust, Go, and JS canonicalizers over the shared corpus
// (spec/corpus/canon.json).
//
// A certificate is a PHP associative array. Keys sort by UTF-16 code unit (the
// spec's rule); values may only be strings, non-negative integers, or nested
// objects (associative arrays). JSON arrays, floats, booleans, and null are
// rejected. Strings are emitted as raw UTF-8 with only the mandated escapes.
final class Canon
{
    public static function canonicalize(array $cert): string
    {
        unset($cert['sig']);
        return self::encodeObject($cert);
    }

    private static function encodeObject(array $obj): string
    {
        uksort($obj, static fn($a, $b) => strcmp(self::utf16($a), self::utf16($b)));
        $parts = [];
        foreach ($obj as $k => $v) {
            $parts[] = self::encodeString((string) $k) . ':' . self::encodeValue($v);
        }
        return '{' . implode(',', $parts) . '}';
    }

    private static function encodeValue($v): string
    {
        if (is_string($v)) {
            return self::encodeString($v);
        }
        if (is_int($v)) {
            if ($v < 0) {
                throw new \RuntimeException("canon: negative integer $v");
            }
            return (string) $v;
        }
        if (is_array($v)) {
            if (self::isList($v)) {
                throw new \RuntimeException('canon: arrays are not permitted in a certificate');
            }
            return self::encodeObject($v);
        }
        $t = is_float($v) ? 'non-integer number' : gettype($v);
        throw new \RuntimeException("canon: $t not permitted in a certificate");
    }

    private static function encodeString(string $s): string
    {
        $out = '"';
        $len = strlen($s);
        for ($i = 0; $i < $len; $i++) {
            $c = ord($s[$i]);
            switch ($c) {
                case 0x22: $out .= '\\"'; break;
                case 0x5c: $out .= '\\\\'; break;
                case 0x08: $out .= '\\b'; break;
                case 0x09: $out .= '\\t'; break;
                case 0x0a: $out .= '\\n'; break;
                case 0x0c: $out .= '\\f'; break;
                case 0x0d: $out .= '\\r'; break;
                default:
                    if ($c < 0x20) {
                        $out .= sprintf('\\u%04x', $c);
                    } else {
                        $out .= $s[$i]; // raw byte: ASCII or UTF-8 continuation
                    }
            }
        }
        return $out . '"';
    }

    // A non-empty array with sequential 0..n-1 integer keys is a JSON array; an
    // empty array is treated as an empty object.
    private static function isList(array $a): bool
    {
        return $a !== [] && array_is_list($a);
    }

    // UTF-8 string to UTF-16BE bytes, so strcmp compares by UTF-16 code unit.
    private static function utf16(string $s): string
    {
        $out = '';
        $len = strlen($s);
        $i = 0;
        while ($i < $len) {
            $c = ord($s[$i]);
            if ($c < 0x80) {
                $cp = $c;
                $i += 1;
            } elseif ($c < 0xE0) {
                $cp = (($c & 0x1F) << 6) | (ord($s[$i + 1]) & 0x3F);
                $i += 2;
            } elseif ($c < 0xF0) {
                $cp = (($c & 0x0F) << 12) | ((ord($s[$i + 1]) & 0x3F) << 6) | (ord($s[$i + 2]) & 0x3F);
                $i += 3;
            } else {
                $cp = (($c & 0x07) << 18) | ((ord($s[$i + 1]) & 0x3F) << 12)
                    | ((ord($s[$i + 2]) & 0x3F) << 6) | (ord($s[$i + 3]) & 0x3F);
                $i += 4;
            }
            if ($cp < 0x10000) {
                $out .= pack('n', $cp);
            } else {
                $cp -= 0x10000;
                $out .= pack('n', 0xD800 + ($cp >> 10)) . pack('n', 0xDC00 + ($cp & 0x3FF));
            }
        }
        return $out;
    }
}

File: go/canon/canon.go
// Package canon implements the BoneMesh restricted-JCS certificate
// canonicalization (security.md §11.1): the exact byte string the mesh root
// signs. Byte-for-byte identical to the Java, Elixir, and Rust implementations
// (and the Go conformance runner) over the shared corpus (spec/corpus/canon.json).
package canon

import (
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"unicode/utf16"
)

// Canonicalize returns the canonical UTF-8 bytes for a certificate object,
// after removing any "sig" member. The value must decode with UseNumber so
// integers survive exactly.
func Canonicalize(cert map[string]any) (string, error) {
	filtered := make(map[string]any, len(cert))
	for k, v := range cert {
		if k != "sig" {
			filtered[k] = v
		}
	}
	var b strings.Builder
	if err := encodeObject(&b, filtered); err != nil {
		return "", err
	}
	return b.String(), nil
}

func encodeObject(b *strings.Builder, m map[string]any) error {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return lessUTF16(keys[i], keys[j]) })
	b.WriteByte('{')
	for i, k := range keys {
		if i > 0 {
			b.WriteByte(',')
		}
		encodeString(b, k)
		b.WriteByte(':')
		if err := encodeValue(b, m[k]); err != nil {
			return err
		}
	}
	b.WriteByte('}')
	return nil
}

func encodeValue(b *strings.Builder, v any) error {
	switch t := v.(type) {
	case string:
		encodeString(b, t)
		return nil
	case json.Number:
		i, err := strconv.ParseInt(t.String(), 10, 64)
		if err != nil {
			return fmt.Errorf("canon: %q is not an integer", t.String())
		}
		if i < 0 {
			return fmt.Errorf("canon: negative integer %d", i)
		}
		b.WriteString(strconv.FormatInt(i, 10))
		return nil
	case map[string]any:
		return encodeObject(b, t)
	default:
		return fmt.Errorf("canon: value type %T not permitted in a certificate", v)
	}
}

func encodeString(b *strings.Builder, s string) {
	b.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"':
			b.WriteString(`\"`)
		case '\\':
			b.WriteString(`\\`)
		case '\b':
			b.WriteString(`\b`)
		case '\t':
			b.WriteString(`\t`)
		case '\n':
			b.WriteString(`\n`)
		case '\f':
			b.WriteString(`\f`)
		case '\r':
			b.WriteString(`\r`)
		default:
			if r < 0x20 {
				fmt.Fprintf(b, `\u%04x`, r)
			} else {
				b.WriteRune(r)
			}
		}
	}
	b.WriteByte('"')
}

func lessUTF16(a, c string) bool {
	ua, uc := utf16.Encode([]rune(a)), utf16.Encode([]rune(c))
	for i := 0; i < len(ua) && i < len(uc); i++ {
		if ua[i] != uc[i] {
			return ua[i] < uc[i]
		}
	}
	return len(ua) < len(uc)
}
